### PR TITLE
feat(issues): add support for jira cloud [WIP]

### PR DESCRIPTION
**Pull request by setchy**
## Changes

_Note: This is a Draft PR.  I am raising it early so that I can begin to get broader feedback and input._

Introduces issues module and adds support for Jira Cloud Issues. 

Work still pending:

- [ ] Unit testing
- [ ] Update document on how to use, limitations (as mentioned above) and make clear this feature is experimental (until it's had time to be hardened etc)
- [ ] Create and maintain sub-task(s) linked to the `Dependency Dashboard` for any open PRs so they can be assigned and planned into team sprint commitments.  This may be best as a separate follow-up PR as it adds net-new functionality.
- [ ] Other things I'm forgetting 😅 
 

## Context

- ref  https://github.com/renovatebot/renovate/issues/20568

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by setchy** [comment]
so the `yarn type-check` step is not happy... 

adding `skipLibCheck: true` to `tsconfig.json` makes it pass, but I'm not sure if that's "ok"

**Comment by viceice** [comment]
it doesn't resolve the mentioned issue. the issue is about using Jira for bitbucket server, not bitbucket Cloud

File: lib/modules/datasource/bitbucket-tags/index.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { cache } from '../../../util/cache/package/decorator';
 import { BitbucketHttp } from '../../../util/http/bitbucket';
 import { ensureTrailingSlash } from '../../../util/url';
+import type { PagedResult, RepoInfoBody } from '../../platform/bitbucket/types';
```
Comment (viceice) [inline lib/modules/datasource/bitbucket-tags/index.ts:4]: don't make changes to unrelated code, open separate PR's if needed

File: lib/modules/platform/bitbucket/utils.ts [hunk 2]
```diff
@@ -1,56 +1,27 @@
 import url from 'url';
+import { defaultSchema } from '@atlaskit/adf-schema/schema-default';
+import {
+  JSONDocNode,
+  JSONTransformer,
+} from '@atlaskit/editor-json-transformer';
```
Comment (viceice) [inline lib/modules/platform/bitbucket/utils.ts:6]: please don't add new deps for functionality which is already provided by other existing deps.

checkout remark

File: tsconfig.json [hunk 3]
```diff
@@ -20,7 +20,8 @@
     "types": ["node", "jest-extended", "expect-more-jest"],
     "allowJs": true,
     "checkJs": true,
-    "importsNotUsedAsValues": "error"
+    "importsNotUsedAsValues": "error",
+    "skipLibCheck": true
```
Comment (viceice) [inline tsconfig.json:24]: no

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by setchy** [comment]
> it doesn't resolve the mentioned issue. the issue is about using Jira for bitbucket server, not bitbucket Cloud

updated to https://github.com/renovatebot/renovate/issues/20568

File: lib/modules/platform/bitbucket/utils.ts (hunk 2 above)
Comment (setchy) [reply lib/modules/platform/bitbucket/utils.ts:6]: I had already looked through the [available remark plugins](https://github.com/remarkjs/remark/blob/main/doc/plugins.md#list-of-plugins) but there isn't any that support converting into [Atlassian Document Format](https://developer.atlassian.com/cloud/jira/platform/apis/document/structure/), which is an Atlassian extension of [prosemirror](https://prosemirror.net/) for rich text fields.  

See [JRACLOUD-77436](https://jira.atlassian.com/browse/JRACLOUD-77436) for current "best" approach to convert markdown -> ADF as recommended by Atlassian

File: lib/modules/datasource/bitbucket-tags/index.ts (hunk 1 above)
Comment (setchy) [reply lib/modules/datasource/bitbucket-tags/index.ts:4]: Sure, I can move this to a separate PR if you really prefer.  I figured after adding a dozen or so new types for Jira Cloud that this would also make sense together.

File: lib/modules/platform/bitbucket/utils.ts (hunk 2 above)
Comment (setchy) [reply lib/modules/platform/bitbucket/utils.ts:6]: That said, I'll play around with downgrading to the v2 Jira Cloud API to see if I can get sufficient rich text formatting for the description/body via markdown 🤞

File: lib/modules/datasource/bitbucket-tags/index.ts (hunk 1 above)
Comment (setchy) [reply lib/modules/datasource/bitbucket-tags/index.ts:4]: ref https://github.com/renovatebot/renovate/pull/20570

File: lib/modules/platform/bitbucket/utils.ts (hunk 2 above)
Comment (setchy) [reply lib/modules/platform/bitbucket/utils.ts]: So I went down the Jira v2 API route, and have it functionally working without any additional packages/deps, but the rendering of issue body was pretty limited as it used the much dated [Atlassian Text Formating Notation](https://jira.atlassian.com/secure/WikiRendererHelpAction.jspa?section=all).

I then returned back to the Jira v3 API and am using remark to get the Github Markdown AST and have implemented a custom AST parser to generate the Atlassian Document Format needed.  More custom code, but at least gets us away from needing the atlaskit packages to do this automagically.

File: tsconfig.json (hunk 3 above)
Comment (setchy) [reply tsconfig.json]: This has been reverted as I no longer am using the atlaskit packages

File: lib/util/http/jira.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import type { HttpOptions, HttpResponse, InternalHttpOptions } from './types';
```
Comment (viceice) [inline lib/util/http/jira.ts:1]: can we extract this to be merged first, even it's not yet used then.

**Review by viceice — CHANGES_REQUESTED** [review]
please create a refactor PR first to add modules/issues/index, which provides the issue API. all calls can be forwarded to platform for first step.

the change all related code to call the issue API instead of platform directly.

File: lib/config/options/index.ts [hunk 5]
```diff
@@ -648,6 +649,13 @@ const options: RenovateOptions[] = [
     default: true,
     supportedManagers: ['npm', 'composer'],
   },
+  {
+    name: 'issueCollector',
```
Comment (setchy) [inline lib/config/options/index.ts:653]: Any recommendations on the new global configuration name to specify the issue collector?  Also thought about `issuePlatform`

File: lib/modules/issues/jira/index.ts [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+import type {
+  JiraIssue,
+  JiraSearchResponse,
+  JiraTransitionsResponse,
+} from './types';
+import * as utils from './utils';
+
+const jiraHttp = new JiraHttp();
+
+export async function findIssue(
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/modules/issues/jira/index.ts]: Done

File: lib/modules/issues/jira/types.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+export interface JiraTransitionsResponse {
+  transitions: JiraTransition[];
+}
+
+export const StatusCategoryKey = [
+  'new',
+  'indeterminate',
+  'in-progress',
+  'done',
+] as const;
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/modules/issues/jira/types.ts]: Done

File: lib/modules/issues/jira/utils.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  return atlassianDocumentFormat;
+}
+
+export function convertAtlassianDocumentFormatToMarkdown(
+  description: AtlassianDocumentFormat
+): string {
+  // const markdown = remark().use(github).process(description);
+
+  /**
+   * TODO: This is a naive implementation where we don't try to convert ADF back to markdown,
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/modules/issues/jira/utils.ts]: That could certainly work - but that custom multiline text field would need to be setup per project which would be complicated at scale.

I've been experimenting with this branch and the naive implementation seems to be ok.  the dashboard jira issue is updated on each run, which isn't such a bad thing imho.  the last updated timestamp is a good reflection of how "current" the dashboard contents are :)

File: lib/modules/platform/bitbucket/index.ts [hunk 9]
```diff
@@ -439,9 +475,46 @@ export async function setBranchStatus({
...
+      'Both bitbucket issues and jira issues are configured.  Please only enable one per repository.'
+    );
+    return null;
+  }
+
+  if (config.hasJiraProjectLinked) {
+    return await jiraIssues.findIssue(title, config.jiraProjectKey);
+  }
+
+  return await findBitbucketIssue(title);
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/modules/platform/bitbucket/index.ts]: Have moved all the code to go through the modules/issues, which as suggested with fallback to platform issues if the type isn't jira.

File: lib/util/http/jira.ts (hunk 4 above)
Comment (setchy) [reply lib/util/http/jira.ts:1]: Certainly - there is lots in this WIP branch that will need to be teased out and added in chunks

https://github.com/renovatebot/renovate/pull/21056 for the jira http util

File: lib/modules/issue/jira/index.ts [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+      .replace(regEx(/\*\*/g), '') // Remove closing bold tags
+      .replace(regEx(/WARN:/g), '⚠️'); // WARN to use emoji
+  }
+}
+
+/**
+ * TODO - What is the best way to retrieve these per repository to use in the Jira Issue Collector?
+ * Retrieving this may vary between platforms(ie: Bitbucket repositories can be natively linked)
+ */
+function getProjectKey(): string {
```
Comment (setchy) [inline lib/modules/issue/jira/index.ts:243]: I'm needing some advice on how best to configure the following for the Jira Issue Collector
-  jira project key for the repository.  
  - for bitbucket cloud, this can be fetched via a call to the /repositories/{repo}/jira/projects endpoint.  how would this value be passed down to the issue collector logic?
  - for other platform types, how would this work?
  - 
- jira cloud url.
  - similarly, this can be fetched from the bitbucket /jira/projects endpoint. 
  - for other platform types, how would this work?

File: lib/modules/issue/index.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    return _issueCollector[prop];
+  },
+};
+export const issueCollector = new Proxy<IssueCollectorAPI>({} as any, handler);
+
+export function initIssueCollector(config: AllConfig): void {
+  setIssueCollectorApi(config);
+}
+
+function setIssueCollectorApi(config: AllConfig): void {
```
Comment (setchy) [inline lib/modules/issue/index.ts:32]: Will use the configured issueCollector, or, fallback to the platform-type issues

File: lib/modules/issue/issue-collector.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import type {
+  EnsureIssueConfig,
+  EnsureIssueResult,
+  Issue,
+} from '../platform/types';
+import type { IssueCollectorAPI } from './types';
+
+export type IssueCollectorId = 'jira';
+
+export abstract class IssueCollector implements IssueCollectorAPI {
```
[the remark this answers is not in the document]
Comment (setchy) [reply lib/modules/issue/issue-collector.ts:12]: A believe there are a handful of situations, but mainly Dependency dashboards, config errors etc at the moment...

I raised https://github.com/renovatebot/renovate/issues/20574 recently to capture the enhancement idea of supporting Confluence spaces - this seems to be similar to what you've done with your local patch @kwelch.

Very much open to better naming suggestions than issue collector. 

Furthermore, for Jira Cloud, I'm interest to go beyond just a single dependency dashboard ticket and generate linked sub-tasks for each of the open PRs,

[the remark this answers is not in the document]
Comment (kwelch) [reply lib/modules/issue/issue-collector.ts:12]: Thanks for the context. I will see if I can spend some time on this during our upcoming tech debt week.

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/issue/issue-collector.ts:12]: this is also used for the config error issue

**Comment by rarkins** [comment]
Closing due to inactivity - trying to keep stale open PRs to a minimum as it's a negative sign to those browsing repos. This topic is welcome to be resumed at any time
